Add utils to inspect fp16/int8 packed weights

csummersea opened the pull request:
Summary:
Since we use prepacked weights in the Fp16 FCs and future Int8 FCs in production Ads models, we provide the python utils to inspect the unpacked format of the weights for debugging purpose. The main interfaces are the following:

```
from deeplearning.numeric_suite.toolkit import packed_weights_inspector
# inspect fp16 packed weights
unpacked_fp16_weights = packed_weights_inspector.extract_fp16_fc_packed_weights(fp16_weight_blob_name)

# inspect int8 packed weights
unpacked_int8_weights, qparams = packed_weights_inspector.extract_int8_fc_packed_weights(int8_weight_blob_name)
```

Test Plan:
```
buck test mode/opt deeplearning/numeric_suite/toolkit/test:packed_weights_inspector_test
```

Differential Revision: D19724474

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 277ff12a:


* **5/5** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=277ff12a9deebed334b54231dbf4bc62d756671e).

### :detective: 4 new failures recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4411283) pytorch_linux_xenial_py3_clang5_asan_build (1/4)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=95113050) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=95113050))
<details>
<summary>
<code>Feb 04 17:58:23 caused by: Connection refused (os error 111)</code>
</summary>

```
Feb 04 17:58:23 +++ eval 'extract_trap_cmd ' 
Feb 04 17:58:23 ++++ extract_trap_cmd 
Feb 04 17:58:23 ++++ printf '%s\n' '' 
Feb 04 17:58:23 +++ printf '%s\n' cleanup 
Feb 04 17:58:23 ++ trap -- ' 
Feb 04 17:58:23 cleanup' EXIT 
Feb 04 17:58:23 ++ which sccache 
Feb 04 17:58:23 ++ sccache --stop-server 
Feb 04 17:58:23 Stopping sccache server... 
Feb 04 17:58:23 error: couldn't connect to server 
Feb 04 17:58:23 caused by: Connection refused (os error 111) 
Feb 04 17:58:23 ++ true 
Feb 04 17:58:23 ++ rm /var/lib/jenkins/sccache_error.log 
Feb 04 17:58:23 rm: cannot remove '/var/lib/jenkins/sccache_error.log': No such file or directory 
Feb 04 17:58:23 ++ true 
Feb 04 17:58:23 ++ SCCACHE_ERROR_LOG=/var/lib/jenkins/sccache_error.log 
Feb 04 17:58:23 ++ SCCACHE_IDLE_TIMEOUT=1200 
Feb 04 17:58:23 ++ RUST_LOG=sccache::server=error 
Feb 04 17:58:23 ++ sccache --start-server 
Feb 04 17:58:23 Starting sccache server... 
Feb 04 17:58:23 ++ sccache --zero-stats 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4411270) pytorch_linux_xenial_py2_7_9_build (2/4)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=95113094) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=95113094))
<details>
<summary>
<code>Feb 04 17:57:56 caused by: Connection refused (os error 111)</code>
</summary>

```
Feb 04 17:57:56 +++ eval 'extract_trap_cmd ' 
Feb 04 17:57:56 ++++ extract_trap_cmd 
Feb 04 17:57:56 ++++ printf '%s\n' '' 
Feb 04 17:57:56 +++ printf '%s\n' cleanup 
Feb 04 17:57:56 ++ trap -- ' 
Feb 04 17:57:56 cleanup' EXIT 
Feb 04 17:57:56 ++ which sccache 
Feb 04 17:57:56 ++ sccache --stop-server 
Feb 04 17:57:56 Stopping sccache server... 
Feb 04 17:57:56 error: couldn't connect to server 
Feb 04 17:57:56 caused by: Connection refused (os error 111) 
Feb 04 17:57:56 ++ true 
Feb 04 17:57:56 ++ rm /var/lib/jenkins/sccache_error.log 
Feb 04 17:57:56 rm: cannot remove '/var/lib/jenkins/sccache_error.log': No such file or directory 
Feb 04 17:57:56 ++ true 
Feb 04 17:57:56 ++ SCCACHE_ERROR_LOG=/var/lib/jenkins/sccache_error.log 
Feb 04 17:57:56 ++ SCCACHE_IDLE_TIMEOUT=1200 
Feb 04 17:57:56 ++ RUST_LOG=sccache::server=error 
Feb 04 17:57:56 ++ sccache --start-server 
Feb 04 17:57:56 Starting sccache server... 
Feb 04 17:57:56 ++ sccache --zero-stats 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4411287) pytorch_macos_10_13_py3_build (3/4)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=95113127) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=95113127))
<details>
<summary>
<code>Feb 04 10:03:17 FAILED: caffe2/CMakeFiles/torch_cpu.dir/quantization/server/fbgemm_pack_op.cc.o</code>
</summary>

```
Feb 04 10:03:16 ../c10/util/Registry.h:296:3: note: expanded from macro 'C10_REGISTER_CLASS' 
Feb 04 10:03:16   C10_REGISTER_TYPED_CLASS(RegistryName, #key, __VA_ARGS__) 
Feb 04 10:03:16   ^ 
Feb 04 10:03:16 ../c10/util/Registry.h:250:33: note: expanded from macro 'C10_REGISTER_TYPED_CLASS' 
Feb 04 10:03:16       Registerer##RegistryName::DefaultCreator<__VA_ARGS__>,                \ 
Feb 04 10:03:16                                 ^ 
Feb 04 10:03:16 3 warnings generated. 
Feb 04 10:03:17 [2334/3006] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/quantization/server/fbgemm_pack_matrix_cache.cc.o 
Feb 04 10:03:17 [2335/3006] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/quantization/server/elementwise_sum_relu_op.cc.o 
Feb 04 10:03:17 [2336/3006] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/quantization/server/fbgemm_pack_op.cc.o 
Feb 04 10:03:17 FAILED: caffe2/CMakeFiles/torch_cpu.dir/quantization/server/fbgemm_pack_op.cc.o  
AVE_THREAD -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wno-write-strings -Wno-unknown-pragmas -Wno-missing-braces -fvisibility=hidden -O2 -DCAFFE2_BUILD_MAIN_LIB -DASMJIT_STATIC -std=gnu++14 -MD -MT caffe2/CMakeFiles/torch_cpu.dir/quantization/server/fbgemm_pack_op.cc.o -MF caffe2/CMakeFiles/torch_cpu.dir/quantization/server/fbgemm_pack_op.cc.o.d -o caffe2/CMakeFiles/torch_cpu.dir/quantization/server/fbgemm_pack_op.cc.o -c ../caffe2/quantization/server/fbgemm_pack_op.cc 
Feb 04 10:03:17 ../caffe2/quantization/server/fbgemm_pack_op.cc:3:10: fatal error: 'caffe2/caffe2/fb/fbgemm/proto/fbgemm_int8.pb.h' file not found 
Feb 04 10:03:17 #include "caffe2/caffe2/fb/fbgemm/proto/fbgemm_int8.pb.h" 
Feb 04 10:03:17          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
Feb 04 10:03:17 1 error generated. 
Feb 04 10:03:17 [2337/3006] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/quantization/server/fully_connected_dnnlowp_acc16_op.cc.o 
Feb 04 10:03:17 In file included from ../caffe2/quantization/server/fully_connected_dnnlowp_acc16_op.cc:1: 
Feb 04 10:03:17 In file included from ../caffe2/quantization/server/fully_connected_dnnlowp_acc16_op.h:3: 
Feb 04 10:03:17 In file included from ../caffe2/quantization/server/fully_connected_dnnlowp_op.h:5: 
Feb 04 10:03:17 In file included from ../caffe2/quantization/server/dnnlowp_op.h:9: 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4411279) pytorch_xla_linux_xenial_py3_6_clang7_build (4/4)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=95113190) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=95113190))
<details>
<summary>
<code>Feb 04 17:58:31 caused by: Connection refused (os error 111)</code>
</summary>

```
Feb 04 17:58:31 +++ eval 'extract_trap_cmd ' 
Feb 04 17:58:31 ++++ extract_trap_cmd 
Feb 04 17:58:31 ++++ printf '%s\n' '' 
Feb 04 17:58:31 +++ printf '%s\n' cleanup 
Feb 04 17:58:31 ++ trap -- ' 
Feb 04 17:58:31 cleanup' EXIT 
Feb 04 17:58:31 ++ which sccache 
Feb 04 17:58:31 ++ sccache --stop-server 
Feb 04 17:58:31 Stopping sccache server... 
Feb 04 17:58:31 error: couldn't connect to server 
Feb 04 17:58:31 caused by: Connection refused (os error 111) 
Feb 04 17:58:31 ++ true 
Feb 04 17:58:31 ++ rm /var/lib/jenkins/sccache_error.log 
Feb 04 17:58:31 rm: cannot remove '/var/lib/jenkins/sccache_error.log': No such file or directory 
Feb 04 17:58:31 ++ true 
Feb 04 17:58:31 ++ SCCACHE_ERROR_LOG=/var/lib/jenkins/sccache_error.log 
Feb 04 17:58:31 ++ SCCACHE_IDLE_TIMEOUT=1200 
Feb 04 17:58:31 ++ RUST_LOG=sccache::server=error 
Feb 04 17:58:31 ++ sccache --start-server 
Feb 04 17:58:31 Starting sccache server... 
Feb 04 17:58:31 ++ sccache --zero-stats 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

csummersea committed:
Add utils to inspect fp16/int8 packed weights (#32979)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32979

Since we use prepacked weights in the Fp16 FCs and future Int8 FCs in production Ads models, we provide the python utils to inspect the unpacked format of the weights for debugging purpose. The main interfaces are the following:

```
from deeplearning.numeric_suite.toolkit import packed_weights_inspector
# inspect fp16 packed weights
unpacked_fp16_weights = packed_weights_inspector.extract_fp16_fc_packed_weights(fp16_weight_blob_name)

# inspect int8 packed weights
unpacked_int8_weights, qparams = packed_weights_inspector.extract_int8_fc_packed_weights(int8_weight_blob_name)
```

Test Plan:
```
buck test mode/opt deeplearning/numeric_suite/toolkit/test:packed_weights_inspector_test
```

Reviewed By: amylittleyang

Differential Revision: D19724474

fbshipit-source-id: 218f17bb27d66f8bac07b4461f6d815182ba0b90
